Remove the direct dependency on Apache Commons. Issue - #7658

Neuw84 opened the pull request:
The followed approach is to use native Java 8 code when the alternative was clear or replicate/borrow when not following iceberg-core approach..

Flink and Spark projects declare Apache Commons as direct dependency and therefore on those subprojects the dependency is not removed.

Closes #7658

nastra commented on nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:
```diff
@@ -662,15 +681,4 @@ private List<String> filterByExtension(String tablePath, String extension) {
         .collect(Collectors.toList());
   }
-
-  private static String addRecordsToFile(Table table, String filename) throws IOException {
```
why the change to move this from here to the top?

Neuw84 replied:
will review the changes and repush, maybe it was a thing of IntelliJ on reformatting.

nastra commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetAvro.java:
```diff
@@ -41,6 +40,74 @@
 
 class ParquetAvro {
 
+  static final GenericData DEFAULT_MODEL =
```
it's hard to follow which changes were exactly done in this file due to the large diff

nastra commented on hive3/src/main/java/org/apache/iceberg/mr/hive/vector/HiveVectorizedReader.java:
```diff
@@ -21,7 +21,6 @@
 import java.io.IOException;
 import java.util.List;
 import java.util.Map;
-import org.apache.commons.lang3.ArrayUtils;
```
should compilation not fail here?

Neuw84 replied:
I did not removed the libraries from Gradle for the subprojects ( Maven expert but not so much on Gradle), I have been searching but seems that the sub-projects were using the parent declared libraries when needed... 

Will fix this line on next re-push and if you can point me to where I can delete the libraries for subprojects will be great!

nastra replied:
seems to be used here: https://github.com/apache/iceberg/blob/c07f2aabc0a1d02f068ecf1514d2479c0fbdd3b0/hive3/src/main/java/org/apache/iceberg/mr/hive/vector/HiveVectorizedReader.java#L104

Neuw84 replied:
Have changed that code to native Java 8 compliant coe

Neuw84 commented:
Hi @nastra, 

Addressed your comments in the latest push, let me know your thoughts. 

One question, there was an error in the tests: 

`package StandardCharsets does not exist`

I am assuming that we are using Java 1.8 for the project no? Maybe there is some issue on the Github Actions config (although is weird because that package was introduced on 1.7).

nastra commented on gcp/src/test/java/org/apache/iceberg/gcp/gcs/TestUtils.java:
```diff
@@ -0,0 +1,151 @@
+import java.io.Serializable;
+import org.testcontainers.shaded.com.google.common.base.Preconditions;
+
+public class TestUtils {
```
`TestUtils` indicates that this is a class containing tests for a `Utils` class, so maybe rename this to something else. 
Also is this a 1:1 copy of the `TestUtils` in aws?

Neuw84 replied:
The files are not exactly the same and I do not like  ( but personal opinion here) the `test-jarscope` thing between projects.

Open to suggestions/guidance here!

nastra commented on gcp/src/test/java/org/apache/iceberg/gcp/gcs/TestUtils.java:
```diff
@@ -0,0 +1,151 @@
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import org.testcontainers.shaded.com.google.common.base.Preconditions;
```
wrong import, should be `org.apache.iceberg.relocated.com.google....`

nastra commented on hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:
```diff
@@ -21,7 +21,6 @@
 import java.nio.charset.StandardCharsets;
 import java.sql.Date;
 import java.sql.Timestamp;
-import org.apache.commons.lang3.StringUtils;
```
seems to be used here: https://github.com/apache/iceberg/blob/c07f2aabc0a1d02f068ecf1514d2479c0fbdd3b0/hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java#L70

Neuw84 replied:
Have changed that code to native Java 8 compliant code

nastra commented on dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java:
```diff
@@ -20,18 +20,25 @@
 
 import com.emc.object.Range;
 import com.emc.object.s3.S3ObjectMetadata;
-import com.emc.object.shadow.org.apache.commons.codec.digest.DigestUtils;
```
isn't that using a shaded version, meaning that we wouldn't have to change this file?

Neuw84 replied:
I replaced the code for native Java 8 code, then that shaded version is not going to be needed.

Neuw84 replied:
I have replaced with native Java code

Neuw84 commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockLocalController.java:
```diff
@@ -276,4 +277,246 @@ public void setMessage(String message) {
+   *
+   * <p>This code is borrowed from `org.apache.commons:commons-io`
+   */
+  public class BoundedInputStream extends FilterInputStream {
```
As it is test code that is not going to be shipped I just though that this was good enough. Open to suggestions here ( it is not very clear why the code is using `BoundedInputStream` and I did not wanted to mess with it. 

Open to suggestions/guidance here.

Neuw84 commented:
Hi @nastra, 

Have addressed your comments and applied spotless everywhere. 

About the open questions let me know what your feelings are.

nastra commented on aws/src/test/java/org/apache/iceberg/aws/s3/TestCommonsS3Utils.java:
```diff
@@ -0,0 +1,199 @@
+import java.io.Serializable;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public class TestCommonsS3Utils {
```
can we remove the `Test` prefix here and the `Test` suffix in `GCSCommonUtilsTest` as `Test` indicates that it contains tests. Why not just `IOUtils` like it was named before?

Neuw84 replied:
Done!

nastra commented on hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:
```diff
@@ -67,12 +66,13 @@ public static MapWork findMapWork(JobConf job) {
-      // So, we don't use the below code that would get the correct MapWork. See HIVE-16985.
-      return null;
+    if (prefixes != null) {
+      if (prefixes.trim().isEmpty()) {
```
is this correct?

Neuw84 replied:
Thanks for the catch. Fixed

Neuw84 commented on dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java:
```diff
@@ -70,7 +72,15 @@ public InputStream createInputStream(Range range) {
+    }
+    md.update(content);
+    byte[] digest = md.digest();
+    metadata.setETag(DatatypeConverter.printHexBinary(digest));
```
Agreed, that is not a long term solution as for Java 9 onwards we are going to need another library. 

Fixing with native code.

nastra commented on dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java:
```diff
@@ -70,9 +71,29 @@ public InputStream createInputStream(Range range) {
     return metadata;
   }
+
+  private static final char[] HEX_ARRAY = "0123456789ABCDEF".toCharArray();
```
`static final` fields should generally be placed at the top of the java file

nastra commented on delta-lake/src/integration/java/org/apache/iceberg/delta/TestSnapshotDeltaLakeTable.java:
```diff
@@ -532,13 +533,14 @@ private void addExternalDatafiles(
       }
-    } catch (DecoderException e) {
+    } catch (UnsupportedEncodingException e) {
+      // In Java 10+ this will not need try-catch
```
is that comment relevant? I think we can probably remove it

Neuw84 replied:
Yes, I suppose that the tooling will catch this when the project gets upgraded to a newer Java version. 

Removed

nastra commented on gcp/src/test/java/org/apache/iceberg/gcp/gcs/GCSCommonUtils.java:
```diff
@@ -0,0 +1,144 @@
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public class GCSCommonUtils {
```
One file is called `CommonsS3...` whereas this one is called `GCSCommons...`. I'd suggest to put the `GCS` / `S3` part at the same position in the name

Neuw84 replied:
done!

nastra commented on hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:
```diff
@@ -67,12 +66,13 @@ public static MapWork findMapWork(JobConf job) {
-      // So, we don't use the below code that would get the correct MapWork. See HIVE-16985.
-      return null;
+    if (prefixes != null) {
+      if (!prefixes.trim().isEmpty()) {
```
both ifs can be folded into a single if like in the previous version of the code

Neuw84 replied:
agreed, I have tested all the combinations just to be sure in a JShell 👍 . 

`import org.apache.commons.lang3.StringUtils;
String a=null;

if (a != null && !StringUtils.isBlank(a)) {
    System.out.println("nullllll check");
    }
if (a != null && !a.trim().isEmpty()) {
        System.out.println("null check native");
    }
a="";
if (a != null && !StringUtils.isBlank(a)) {
        System.out.println("empty check ");
    }
if (a != null && !a.trim().isEmpty()) {
        System.out.println("empty check native");
    }
a=" ";
if (a != null && !StringUtils.isBlank(a)) {
        System.out.println("blank check");
    }
if (a != null && !a.trim().isEmpty()) {
        System.out.println("blank check native");
    }
a="bob";
if (a != null && !StringUtils.isBlank(a)) {
        System.out.println("things check");
    }
    if (a != null && !a.trim().isEmpty()) {
        System.out.println("things check native");
    }
a="  bob  ";
if (a != null && !StringUtils.isBlank(a)) {
        System.out.println("trim things check");
    }
    if (a != null && !a.trim().isEmpty()) {
        System.out.println("trim things check native");
    }`

nastra commented on hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:
```diff
@@ -67,10 +66,11 @@ public static MapWork findMapWork(JobConf job) {
+        // Currently SMB is broken, so we cannot check if it's  compatible with IO elevator.
+        // So, we don't use the below code that would get the correct MapWork. See HIVE-16985.
+        return null;
+      }
```
I think } is an extra one and needs to be removed?

Neuw84 replied:
yep you are right, seems to be weird because yesterday I compiled everything.

nastra reviewed (commented):
please run `./gradlew spotlessApply` to fix formatting

nastra commented on hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:
```diff
@@ -67,10 +66,10 @@ public static MapWork findMapWork(JobConf job) {
-      // So, we don't use the below code that would get the correct MapWork. See HIVE-16985.
-      return null;
+    if (prefixes != null && !prefixes.trim().isEmpty()) {
+        // Currently SMB is broken, so we cannot check if it's  compatible with IO elevator.
```
formatting seems to be off?

Neuw84 replied:
yes, fixed on latest commit. 

If you can retry the CI/CD pipeline it will be great.

Thanks!

nastra reviewed (commented):
changes LGTM, but I think we also need to update `TestS3InputStream` / `TestS3FileIOIntegration` / `ManifestReadBenchmark`

Neuw84 commented:
Done with the required changes. 

For `ManifestReadBenchmark` I have also changed the `Files.createTemp` from Guava as it collides with `java.nio.File.Files` and the change was small.

Neuw84 commented:
Hi @nastra, @Fokko, 

I have merged the changes from this pull request ( some tests changed)

https://github.com/apache/iceberg/pull/7996

Let me know whether I need to change anything!

nastra commented:
@Neuw84 could you please rebase against latest master and squash all commits together?

Neuw84 commented:
> @Neuw84 could you please rebase against latest master and squash all commits together?

Done @nastra!
